File: morning_exercises/day14.cs
public class BSTNode3(int val = 0, BSTNode3? left = null, BSTNode3? right = null)
{
    public int val = val;
    public BSTNode3? left = left;
    public BSTNode3? right = right;

    public static void InOrderLA(BSTNode3? root, List<int> result, int k)
    {
        if (root==null) return;

        InOrderLA(root.left, result, k);
        result.Add(root.val);
        InOrderLA(root.right, result, k);
    }

    public static int kthSmallestLA(BSTNode3? root, int k)
    {
        List<int> result = [];
        InOrderLA(root, result, k);
        return result[k-1];
    }

    public static void InOrder(BSTNode3? root, List<int> result, List<int> answer, int k)
    {
        if (root==null || answer[0]!=int.MinValue) return;

        InOrder(root.left, result, answer, k);
        result[0]++;
        if (result[0]==k)
        {
            answer[0]=root.val;
            return;
        }
        InOrder(root.right, result, answer, k);
    }

    public static int kthSmallest(BSTNode3? root, int k)
    {
        List<int> result = [0];
        List<int> answer = [int.MinValue];

        InOrder(root, result, answer, k);

        return answer[0];
    }
}

File: morning_exercises/day14.ts
class BSTNode3 {
    constructor(
        public val: number = 0,
        public left: BSTNode3 | null = null,
        public right: BSTNode3 | null = null
    ) {}

    static inOrderLA(root: BSTNode3 | null, result: number[]): void {
        if (root==null) return;

        BSTNode3.inOrderLA(root.left, result);
        result.push(root.val);
        BSTNode3.inOrderLA(root.right, result);
    }

    static kthSmallestLA(root: BSTNode3 | null, k: number): number {
        let result: number[] = [];
        BSTNode3.inOrderLA(root, result);
        return result[k-1];
    }

    static inOrder(root:BSTNode3 | null, result: number[], answer: number[], k: number): void {
        if (root == null || answer[0]!==-Infinity) return;

        BSTNode3.inOrder(root.left, result, answer, k);
        result[0]++;
        if (k===result[0]) {
            answer[0]=root.val;
            return;
        }
        BSTNode3.inOrder(root.right, result, answer, k);
    }

    static kthSmallest(root: BSTNode3 | null, k: number): number {
        let result: number[] = [0];
        let answer: number[] = [-Infinity];

        BSTNode3.inOrder(root, result, answer, k);

        return answer[0];
    }
}

{
    const root = new BSTNode3(6);
    root.left = new BSTNode3(4);
    root.right = new BSTNode3(9);
    root.left.left = new BSTNode3(2);
    root.left.right = new BSTNode3(5);

    console.log(BSTNode3.kthSmallestLA(root,2));
    console.log(BSTNode3.kthSmallest(root,2));
}
